driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,351 @@
+         * <p/>
+         * When a host's error rate overpasses the threshold specified in {@link #withInclusionThreshold}, the host
+         * will be considered excluded, and this for a certain period of time defined by this method. After this period,
+         * a previously excluded host will be included again unless/until it reaches the threshold again.
+         * <p/>
+         * If the error tracker, during an update, detects that a currently excluded host continues to overpass the
+         * error rate threshold, this host's exclusion time will be reset to the start.
+         * <p/>
+         * Default value for the retry period is 120 seconds.
+         *
+         * @param retryPeriod         the period of exclusion for a host.
+         * @param retryPeriodTimeUnit the time unit for the retry period.
+         * @return this {@link Builder} instance, for method chaining.
+         */
+        public Builder withRetryPeriod(long retryPeriod, TimeUnit retryPeriodTimeUnit) {
+            this.retryPeriod = retryPeriodTimeUnit.toMillis(retryPeriod);
+            return this;
+        }
+
+        /**
+         * Defines the clock to be used by the host {@link Meter}s for tracking error rates.
+         *
+         * @param clock the clock to use for tracking error rates.
+         * @return this {@link Builder} instance, for method chaining.
+         */
+        @VisibleForTesting
+        Builder withClock(Clock clock) {
+            this.clock = clock;
+            return this;
+        }
+
+        /**
+         * Creates the {@link ErrorAwarePolicy} instance.
+         *
+         * @return the newly created {@link ErrorAwarePolicy}.
+         */
+        public ErrorAwarePolicy build() {
+            return new ErrorAwarePolicy(this);
+        }
+    }
+
+    class PerHostErrorTracker implements LatencyTracker {
+
+        private final double errorsThreshold;
+
+        private final ConcurrentMap<Host, Meter> hostsMeters = new ConcurrentHashMap<Host, Meter>();
+
+        private final ConcurrentMap<Host, Long> exclusionTimes = new ConcurrentHashMap<Host, Long>();
+
+        private final Clock clock;
+
+        private final ErrorsFilter errorsFilter;
+
+        PerHostErrorTracker(double errorsThreshold, ErrorsFilter errorsFilter, Clock clock) {
+            this.errorsThreshold = errorsThreshold;
+            this.errorsFilter = errorsFilter;
+            this.clock = clock;
+        }
+
+        @Override
+        public void update(Host host, Statement statement, Exception exception, long newLatencyNanos) {
+            if (exception == null) {
+                return;
+            }
+            if (!errorsFilter.shouldConsiderError(exception, host)) {
+                return;
+            }
+            Meter hostMeter = hostsMeters.get(host);
+            if (hostMeter == null) {
+                hostMeter = hostsMeters.putIfAbsent(host, new Meter(clock));
```

Comment:
`putIfAbsent` returns the entry that was already there, but if nothing was there it returns null so we end up with a null Meter.